Add support for PHP code.

chobie opened the pull request:
I'm writing PHP port of flatbuffers, Our main objectives are to communicate 
native client and php server application. 

Ported from: Java, and C#.
Status: Basic feature and test implemented.

[TODO]

* uint, long, ulong support
  * PHP's integer is platform dependant.
    32bit machine and Windows can't use 64bit values.
    I'll plan to add wrapper class or implementing ByteBuffer as
    C-Extension (and use GMP extension for calculating big integer).
    ref: http://php.net/manual/en/language.types.integer.php

* refactor idl_gen_php.cpp
  * now, work in progress. almost case it can generate correct file.

* fix API

* add fuzzy, ByteBuffer test.

After finish my work (probably, it'll take 1 ~ 3 weeks), I'll sign CLA and mention this issue again.

gwvo commented:
I have absolutely zero experience with PHP, but this looks pretty good from what I can tell. A good amount of tests, make sure you also test the utf8 code that is lurking in there somewhere.

Anyone out there that knows PHP want to help code-review?

chobie commented:
Okay, I've finished basic implementation and tests. now ready to review.

information about php implementation: 
https://github.com/google/flatbuffers/commit/e7fb54bc0d50d7fd34ef7f9129edd8f3c07dffb1

I'd like to release this as experimental support PHP as It's easy to introduce
flatbuffers to php users who got involved. 

Our code respects PSR-2 and PSR-4 coding standards
(these are PHP community driven coding standards). It doesn't have big api design issue.

  ref: http://www.php-fig.org/psr/psr-2/ http://www.php-fig.org/psr/psr-4/

## Limitations:

* PHP's integer is platform dependent. we strongly recommend use 64bit machine
  and don't use uint, ulong types as prevent overflow issue.
  ref: http://php.net/manual/en/language.types.integer.php

* php don't support float type. floating point numbers are always parsed as double precision internally.
  ref: http://php.net/manual/en/language.types.float.php

* ByteBuffer is little bit slow implementation due to many chr/ord function calls. Especially encoding objects.
  This is expected performance as PHP5 has parsing arguments overhead. probably we'll add C-extension.

php doc comment is still work in progress. I'll prettify it during experimental phase (probably in this year.) 

Original commits:
https://github.com/chobie/flatbuffers/compare/b6cc8cd7b5ea352a0ceee0e47d466d8038cb0397...8d670f6da7b7eeb8b708a962b585779883f8b4c5

chobie commented:
oops, maybe I've put string but. i'll check this later

gwvo commented on composer.json:
```diff
@@ -0,0 +1,18 @@
+{
```
Does this file have to be in the root? We'd like to keep it as uncluttered as possible.

kornrunner replied:
Unfortunately yes, if we intend to make it available as a [Composer] (https://getcomposer.org/) package via [Packagist] (https://packagist.org/) - de facto standard dependency manager in PHP world.

chobie replied:
Yea, Basically, we have to put `composer.json` at top level of repository.

(little bit off topic) I've tested sub directory version which using github svn support.
it works good on local machine. but packagist detects the url as git :( 

e.g) https://github.com/chobie/composer-test


But now, current PHP implementation is still experimental. we can postpone this decision till stable release.

gwvo replied:
ok, makes sense

gwvo commented:
Ok, if noone else has comments, I'd say it looks good for merge.

PHPUnit tests would be good if that dependency is easily available for whoever wants to be running the tests.

Can you change the -l to --php? I've been meaning to stop the single character option madness, and your commit might as well be the first. I'll change all the other languages shortly, and deprecate the short ones.

rw commented on tests/phpTest.php:
```diff
@@ -0,0 +1,563 @@
+
+    testByteBuffer($assert);
+    fuzzTest1($assert);
+//    testUnicode($assert);
```
This seems useful, why is it commented out?

chobie replied:
You know this test is javascript port, but I can't find `unicode_test.mon` on tests directory.
I'll implement this when I have chance to setup node.js.

rw commented on tests/phpTest.php:
```diff
@@ -0,0 +1,563 @@
+    $short_val = -32222; // 0x8222;
+    $ushort_val = 0xFEEE;
+    $int_val = 0x83333333 | 0;
+    // for now
```
Can PHP not encode these overflowing ints?

chobie replied:
~ int types are okay since PHP's integer type is signed long (platform dependent).

AFAIK, over uint types requires too much implementation when converting it on userland code due to overflowing integer automatically converted as double internally and lacks functions. 
(also I'm not sure how can i encode those value correctly)


One interesting thing, php-mongo-driver has (https://github.com/mongodb/mongo-php-driver/commit/fa414d10189e48c57c926833d111a2412b733778) dropped uint and ulong types as majority users doesn't use them.
We can those values with extension. maybe I'll write it when ByteBuffer API has been fixed.

rw commented on tests/phpTest.php:
```diff
@@ -0,0 +1,563 @@
+        $uut->putInt(2, 0x0A0B0C0D);
+    });
+
+    if (PHP_INT_SIZE > 4) {
```
Is this check for 32 versus 64 bit platforms?

chobie replied:
Yes. specifically it checks `long` size. windows long type is 32 bit on 32/64bit machine.
Usually we use `PHP_INT_SIZE` constant when checking int size.

rw commented on tests/phpTest.php:
```diff
@@ -0,0 +1,563 @@
+    $bb = Google\FlatBuffers\ByteBuffer::wrap($data);
+    test_buffer($assert, $bb);
+
+    // Second, let's create a FlatBuffer from scratch in JavaScript, and test it also.
```
Old mention of Javascript here.

rw commented on php/ByteBuffer.php:
```diff
@@ -0,0 +1,437 @@
+    public function getShort($index)
+    {
+        $result = $this->readLittleEndian($index, 2);
+        $helper = pack("v", $result);
```
Is this a way to make the value signed? A comment or helper function would be good here.

chobie replied:
Yes, same as https://github.com/google/flatbuffers/pull/308#discussion_r44659373 comment.
helper function would be nice. I'll add that soon.

rw commented on php/ByteBuffer.php:
```diff
@@ -0,0 +1,437 @@
+    /**
+     *
+     */
+    public function reset()
```
Should this also reset vtables and other variables?

chobie replied:
nope. this came from Java's Buffer API (http://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html#reset--).
C# implementation also clear pos only. (https://github.com/google/flatbuffers/blob/master/net/FlatBuffers/ByteBuffer.cs#L51)


btw, Java and C#'s FlatBuffersBuilder doesn't have reset API.

https://github.com/google/flatbuffers/blob/master/java/com/google/flatbuffers/FlatBufferBuilder.java#L436
https://github.com/google/flatbuffers/blob/master/net/FlatBuffers/FlatBufferBuilder.cs#L351

I've used C#(Unity) and PHP, For now, I'm following C# implementations.

rw commented on php/ByteBuffer.php:
```diff
@@ -0,0 +1,437 @@
+    /**
+     * @return bool
+     */
+    public static function isLittleEndian()
```
The big endian code path seems untested, is that correct?

chobie replied:
yea, maybe it's okay. codes are same as C#.
https://github.com/google/flatbuffers/blob/master/net/FlatBuffers/ByteBuffer.cs#L102

I'll add test case later.

rw commented on php/FlatbufferBuilder.php:
```diff
@@ -0,0 +1,918 @@
+     */
+    public function endVector()
+    {
+        $this->putInt($this->vector_num_elems);
```
Why is this signed?

chobie replied:
Since I've referenced CSharp and Java implementations.

https://github.com/google/flatbuffers/blob/master/net/FlatBuffers/FlatBufferBuilder.cs#L211
https://github.com/google/flatbuffers/blob/master/java/com/google/flatbuffers/FlatBufferBuilder.java#L253

You are right, these should be uint.

rw commented on php/FlatbufferBuilder.php:
```diff
@@ -0,0 +1,918 @@
+        return $this->offset();
+    }
+
+    protected function is_utf8($bytes)
```
I understand why this is here... is there a library function we can use instead?

rw replied:
Since it's here it should be tested.

chobie replied:
There are mbstring extension and polyfill-mbstring.

http://php.net/manual/en/book.mbstring.php
https://github.com/symfony/polyfill-mbstring

mbstring depends current detecting order (http://php.net/manual/en/function.mb-detect-order.php) when detecting passed string encoding.
Sometimes it returns wrong encoding when order is not good. Especially, Japanese language often returns wrong encoding with `mb_detect_encoding` as it has several encoding and similar pattern.

I'll add utf8 test case later.

rw commented on php/ByteBuffer.php:
```diff
@@ -0,0 +1,437 @@
+     * @param $offset
+     * @param $value
+     */
+    public function putUint($offset, $value)
```
This is the same code as `putInt`. How does it work?

chobie replied:
I agree, It should be add some comments and put error check on `getUint`.

Currently, https://github.com/chobie/flatbuffers/commit/e7fb54bc0d50d7fd34ef7f9129edd8f3c07dffb1#diff-66510972612fc4e230137135eefbc6c8R225 checks overflowing value.

On 32bit machine (and Windows): 
overflowing unsigned integer can't put with this API as design.

On 64bit machine: unsigned integer value are supported natively, so it doesn't need special task when encoding to bytes.
It needs `unpack` function when decoding value.  
https://github.com/chobie/flatbuffers/commit/e7fb54bc0d50d7fd34ef7f9129edd8f3c07dffb1#diff-8bca56a260f1ea2ed9560137c20536d5R376

but I get little bit concerned. I'll make sure this behaviour.

rw commented:
I'd like to see the generated bytes for the example data, and verify that they are exactly the same as the Java bytes. This is done in the Go and Python versions. Is that too strict, @gwvo?

chobie commented:
Thanks reviewing my code. I'll take a look into that when I back home.

gwvo commented:
@rw : it's nice to have, but it is a fragile test, so not required as far as I'm concerned.

rw commented:
@chobie Thanks for replying to my line comments.

I think that users should be prevented from writing values that are too big for the schema datatype. For example, I should not be allowed to write `0xFFFF` into a FlatBuffers field of type `uint8`.

We have checks for this in the Python port. What do you think?

chobie commented:
@gwvo fixed several things, includes -l to --php flags. can you check commits?

> PHPUnit tests would be good if that dependency is easily available for whoever wants to be running the tests.
Okay, I'll switch to PHPUnit when I added travis support. 

@rw Yea, It would be nice, LL users often forgot type systems. I'll take a look into your implementation later.

gwvo commented:
@chobie Thanks! Once you feel you've addressed @rw's concerns, we can merge.

chobie commented:
@gwvo yea, I've implemented that. now, squashed & forced pushed php branch.
travis seems not good at this time, probably their system trouble. ill check travis tomorrow.

original branch is here.
https://github.com/chobie/flatbuffers/compare/master...chobie:php-origin-20151118

I'm planning to do remaining PHP tasks (e.g mutation, json parsing, reflection... etc). before do these, I need to play C++ implementation as I should know more internal implementation.
until then, my main contribution will be C#(Unity) :)

gwvo commented:
Ok, I'll merge, and any improvements to the PHP port can be follow-up PRs.
Thanks for your contribution!

kornrunner commented:
Could you please add package to https://packagist.org?
